File: 1.js
// LeetCode 1: Two Sum
// Reference: https://leetcode.com/problems/two-sum/solutions/234005/neat-javascript-map-o-n/?envType=study-plan-v2&envId=top-interview-150


// Using ES6 data structure Map. Time complexity is O(n).

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let map = new Map();

    for (let i = 0; i < nums.length; i++){
        if (map.has(target - nums[i])){
            return [map.get(target - nums[i]), i];
        } else {
            map.set(nums[i], i);
        }
    }
    return [];
    
};

File: 1.java
// LeetCode 1: Two Sum
// Reference: https://leetcode.com/problems/two-sum/solutions/2671539/java-100-runtime-easy-to-understand/?envType=study-plan-v2&envId=top-interview-150

import java.util.Map;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        int n = nums.length;

        Map<Integer, Integer> map = new HashMap<>();

        int[] result = new int[2];

        for (int i = 0; i < n; i++){
            if (map.containsKey(target - nums[i])){
                result[1] = i;
                result[0] = map.get(target - nums[i]);
                return result;
            }
            map.put(nums[i], i);
        }
        return result;
    }
}